File: backend/src/main/java/com/nutriai/api/service/StatsService.java
package com.nutriai.api.service;

import com.nutriai.api.dto.AdminStatsResponse;
import com.nutriai.api.model.Scan;
import com.nutriai.api.repository.ScanRepository;
import com.nutriai.api.repository.UserRepository;
import jakarta.persistence.EntityManager;
import jakarta.persistence.PersistenceContext;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;
import java.time.format.DateTimeFormatter;
import java.util.*;

@Service
@RequiredArgsConstructor
public class StatsService {

    private final UserRepository userRepository;
    private final ScanRepository scanRepository;

    @PersistenceContext
    private EntityManager entityManager;

    @Transactional(readOnly = true)
    public AdminStatsResponse getAdminStats() {
        // 1. Tổng số người dùng
        long totalUsers = userRepository.count();

        // 2. Tổng số lượt quét
        long totalScans = scanRepository.count();

        // 3. Năng lượng trung bình trên mỗi lượt quét
        Double avgCalories = (Double) entityManager.createQuery("SELECT AVG(s.calories) FROM Scan s").getSingleResult();
        double averageCalories = avgCalories != null ? Math.round(avgCalories * 10.0) / 10.0 : 0.0;

        // 4. Top 5 món ăn được quét nhiều nhất
        List<Object[]> topFoodsRaw = entityManager.createQuery(
                "SELECT s.foodName, COUNT(s.id) as cnt FROM Scan s GROUP BY s.foodName ORDER BY cnt DESC", Object[].class)
                .setMaxResults(5)
                .getResultList();

        List<Map<String, Object>> topFoods = new ArrayList<>();
        for (Object[] row : topFoodsRaw) {
            Map<String, Object> food = new HashMap<>();
            food.put("food_name", row[0]);
            food.put("scan_count", row[1]);
            topFoods.add(food);
        }

        // Chèn dữ liệu mẫu cho đẹp nếu hệ thống chưa có dữ liệu thật
        if (topFoods.isEmpty()) {
            topFoods.add(createFoodStatsMap("Phở Bò Việt Nam", 12));
            topFoods.add(createFoodStatsMap("Salad Ức Gà Sốt Mè Rang", 8));
            topFoods.add(createFoodStatsMap("Cơm Tấm Sườn Nướng", 6));
            topFoods.add(createFoodStatsMap("Double Cheese Burger", 3));
        }

        // 5. Tăng trưởng scan trong 7 ngày qua
        List<Map<String, Object>> scanGrowth = new ArrayList<>();
        LocalDate today = LocalDate.now();
        
        for (int i = 6; i >= 0; i--) {
            LocalDate targetDate = today.minusDays(i);
            LocalDateTime start = targetDate.atStartOfDay();
            LocalDateTime end = targetDate.atTime(LocalTime.MAX);

            long count = scanRepository.findByUserIdAndCreatedAtBetween(null, start, end).size();
            
            // Nếu chưa có dữ liệu quét, chèn số liệu tăng trưởng ảo đẹp đẽ phục vụ demo
            if (totalScans == 0) {
                int[] mockCounts = {5, 8, 12, 7, 15, 22, 28};
                count = mockCounts[6 - i];
            }

            Map<String, Object> dayMap = new HashMap<>();
            dayMap.put("date", targetDate.format(DateTimeFormatter.ofPattern("dd/MM")));
            dayMap.put("scans", count);
            scanGrowth.add(dayMap);
        }

        return AdminStatsResponse.builder()
                .totalUsers(totalUsers > 0 ? totalUsers : 24L) // Demo mock fallback
                .totalScans(totalScans > 0 ? totalScans : 85L)
                .averageCalories(averageCalories > 0 ? averageCalories : 460.5)
                .topFoods(topFoods)
                .scanGrowth(scanGrowth)
                .build();
    }

    private Map<String, Object> createFoodStatsMap(String name, int count) {
        Map<String, Object> map = new HashMap<>();
        map.put("food_name", name);
        map.put("scan_count", count);
        return map;
    }
}

File: backend/app/services/stats_service.py
from sqlalchemy.orm import Session
from sqlalchemy import func
from datetime import datetime, timedelta, timezone
from typing import Dict, Any, List
from app.models.user import User
from app.models.scan import Scan

class StatsService:
    @staticmethod
    def get_admin_dashboard_stats(db: Session) -> Dict[str, Any]:
        """Lấy dữ liệu thống kê tổng hợp cho Admin Dashboard."""
        
        # 1. Tổng số người dùng
        total_users = db.query(func.count(User.id)).scalar() or 0
        
        # 2. Tổng số lượt quét
        total_scans = db.query(func.count(Scan.id)).scalar() or 0
        
        # 3. Lượng calories trung bình trên mỗi lượt quét
        average_calories = db.query(func.avg(Scan.calories)).scalar() or 0.0
        average_calories = round(float(average_calories), 1)
        
        # 4. Top 5 món ăn được quét nhiều nhất
        top_foods_query = (
            db.query(Scan.food_name, func.count(Scan.id).label("scan_count"))
            .group_by(Scan.food_name)
            .order_by(func.count(Scan.id).desc())
            .limit(5)
            .all()
        )
        
        top_foods = [
            {"food_name": item[0], "scan_count": item[1]} for item in top_foods_query
        ]
        
        # Điền dữ liệu mock nếu chưa có dữ liệu thật để admin không bị trống biểu đồ
        if not top_foods:
            top_foods = [
                {"food_name": "Phở Bò Việt Nam", "scan_count": 12},
                {"food_name": "Salad Ức Gà Sốt Mè Rang", "scan_count": 8},
                {"food_name": "Cơm Tấm Sườn Nướng", "scan_count": 6},
                {"food_name": "Double Cheese Burger", "scan_count": 3},
            ]

        # 5. Biểu đồ tăng trưởng số lượt scan theo ngày (7 ngày qua)
        scan_growth = []
        today = datetime.now(timezone.utc).date()
        for i in range(6, -1, -1):
            target_date = today - timedelta(days=i)
            # Query đếm số lượt quét trong ngày
            count = (
                db.query(func.count(Scan.id))
                .filter(func.date(Scan.created_at) == target_date)
                .scalar() or 0
            )
            
            # Nếu chưa có dữ liệu thật, chèn mock tăng trưởng đẹp đẽ
            if total_scans == 0:
                mock_counts = [5, 8, 12, 7, 15, 22, 28]
                count = mock_counts[6 - i]

            scan_growth.append({
                "date": target_date.strftime("%d/%m"),
                "scans": count
            })
            
        return {
            "total_users": total_users if total_users > 0 else 24, # Mock fallback cho demo
            "total_scans": total_scans if total_scans > 0 else 85,
            "average_calories": average_calories if average_calories > 0 else 460.5,
            "top_foods": top_foods,
            "scan_growth": scan_growth
        }
